### PR TITLE
Corrected PulpImport/Export handling of sub-content.

### LINKED ISSUE #2192: PulpImport/Export of kickstart repos with subrepos broken
See https://bugzilla.redhat.com/show_bug.cgi?id=2040870 for details.

### PR DESCRIPTION
**Pull request by ggainey**
fixes #2192.
[nocoverage]

### CODE REVIEW

File: pulpcore/app/modelresource.py [hunk 1]
```diff
@@ -92,9 +100,19 @@ def before_import_row(self, row, **kwargs):
...
-            "content", "relative_path"
-        )
+        vers_content = ContentArtifact.objects.filter(
+            content__in=self.repo_version.content
+        ).order_by("content", "relative_path")
+        if self.content_mapping:
+            all_content = []
+            for c in self.content_mapping.values():
+                all_content.extend(c)
+            mapped_cas = ContentArtifact.objects.filter(content__in=all_content).order_by(
```
Comment (dralley) [inline pulpcore/app/modelresource.py:110]: Can `all_content` not just be a queryset rather than a list?

Comment (ggainey) [reply pulpcore/app/modelresource.py:110]: content_mapping is a {<repo-name>: [list-of-ids]} - there are no queries at this point. I could build a bunch of queries of CA.objects.filter(content__in=self.content_mapping[repo].values()), and union them all - but I don't see how that makes the code any clearer/faster/more-performant?

**Review by dralley — APPROVED** [review]
This looks reasonable, caveat that I have limited background with import/export

File: pulpcore/app/modelresource.py [hunk 2]
```diff
@@ -92,9 +100,19 @@ def before_import_row(self, row, **kwargs):
...
+                "content", "relative_path"
+            )
+            return vers_content.union(mapped_cas)
+        else:
+            return vers_content
```
Comment (mdellweg) [inline pulpcore/app/modelresource.py:115]: If you dislike, please ignore:
So this function is supposed to return a properly sorted queryset on ContentArtifact, right?
How about we use the `if self.content_mapping:` code path to only augment the `vers_content` queryset in place, and move the sorting to the only return statement? Something like [shortened]
```python
vers_content = ContentArtifact.objects.filter(...)
if self.content_mapping:
    for content_ids in self.content_mapping.values():
        vers_content = vers_content.union(ContentArtifact.objects.filter(pk__in=content_ids)
return vers_content.order_by("content", "relative_path")
```

**Review by mdellweg — APPROVED** [review]
I kind of refactored the PR in my head to understand. So please feel free to ignore my comments.

File: CHANGES/2192.bugfix [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Fixed import/export of repositories with sub-content.
+
+An example would be the sub-repositories in pulp_rpm
+DistributionTrees.
+
```
[the remark this answers is not in the document]
Comment (ggainey) [reply CHANGES/2192.bugfix:5]: If a mechanical limit breaks towncrier, it should have ascript that autoenforces it (like black does for code). If the resulting changelog doesn't ccare - then having "shorter" lines shouldn't be a problem. I can change this if it's that important, but it feels like change for-the-sake-of-change.

File: pulpcore/app/modelresource.py [hunk 4]
```diff
@@ -66,11 +66,19 @@ class ContentArtifactResource(QueryModelResource):
     ContentArtifact is different from other import-export entities because it has no 'natural key'
     other than a pulp_id, which aren't shared across instances. We do some magic to link up
     ContentArtifacts to their matching (already-imported) Content.
+
+    Some plugin-models have sub-repositories. We take advantage of the content-mapping
+    machinery to account for those contentartifacts as well.
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/modelresource.py:71]: Either way... I never know what is right there.

File: pulpcore/app/modelresource.py [hunk 5]
```diff
@@ -66,11 +66,19 @@ class ContentArtifactResource(QueryModelResource):
...
     """
 
     artifact = fields.Field(
         column_name="artifact", attribute="artifact", widget=ForeignKeyWidget(Artifact, "sha256")
     )
+    content_mapping = None
```
[the remark this answers is not in the document]
Comment (ggainey) [reply pulpcore/app/modelresource.py:77]: Yes? We pass in the content-mapping at creation time. setup-query needs it to do the work that need to happen.

File: pulpcore/app/modelresource.py (hunk 2 above)
Comment (ggainey) [reply pulpcore/app/modelresource.py:115]: Sure, that looks much cleaner/more-readable, will do

File: pulpcore/app/modelresource.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/modelresource.py:77]: For that it is set as an instance variable by the `__init__` function.

[the remark this answers is not in the document]
Comment (ggainey) [reply pulpcore/app/modelresource.py:77]: Ah, good point, duh - changing

File: pulpcore/app/modelresource.py (hunk 2 above)
Comment (dralley) [reply pulpcore/app/modelresource.py:115]: +1

**Review by mdellweg — APPROVED** [review]

**Comment by ggainey** [comment]
> CI Failure is intermittent and unrelated. Can we merge anyway?

I can't - but if someone else has button-access, I'm good w/ pushing it.

**Review by dkliban — APPROVED** [review]

**Pull request merged**